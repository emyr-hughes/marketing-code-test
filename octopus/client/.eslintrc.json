{
  "env": {
    "browser": true,
    "es2021": true
  },
  "extends": [
    "plugin:react/recommended",
    "airbnb",
    "prettier",
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended"
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true
    },
    "ecmaVersion": 2020,
    "sourceType": "module"
  },
  "plugins": ["react", "@typescript-eslint"],
  "settings": {
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx", ".json"],
        "paths": ["src"]
      }
    }
  },
  "rules": {
    "quotes": ["error", "double"],
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/extensions": [
      "error",
      "ignorePackages",
      {
        "js": "never",
        "mjs": "never",
        "jsx": "never",
        "ts": "never",
        "tsx": "never"
      }
    ],
    "import/prefer-default-export": "off",
    "react/jsx-uses-react": "error",
    "react/react-in-jsx-scope": "OFF",
    "react/jsx-uses-vars": "error",
    "react/jsx-props-no-spreading": "off",
    "react/require-default-props": "off",
    "camelcase": "off",
    "no-param-reassign": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "no-plusplus": "off",
    "spaced-comment": "off",
    "arrow-body-style": "off",
    "no-nested-ternary": "off",
    "global-require": "off",
    "react/jsx-filename-extension": [
      2,
      { "extensions": [".js", ".jsx", ".ts", ".tsx"] }
    ]
  }
}
