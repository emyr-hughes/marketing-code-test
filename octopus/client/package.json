{
  "name": "octopus-client",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@apollo/client": "^3.4.16",
    "@apollo/react-hooks": "^3.1.3",
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/react": "12.1.2",
    "@testing-library/user-event": "^7.1.2",
    "graphql": "^14.6.0",
    "react": "^16.12.0",
    "react-dom": "^16.12.0",
    "react-router-dom": "^5.1.2",
    "react-scripts": "3.4.4",
    "styled-components": "^5.3.3",
    "typescript": "4.1.2"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jest-environment-jsdom-sixteen",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@types/jest": "^27.0.2",
    "@types/node": "^16.11.6",
    "@types/react": "^17.0.34",
    "@types/react-dom": "^17.0.11",
    "@types/styled-components": "^5.1.15",
    "@typescript-eslint/eslint-plugin": "4.0.1",
    "@typescript-eslint/parser": "4.0.1",
    "eslint-config-airbnb": "18.2.1",
    "eslint-config-prettier": "8.3.0",
    "eslint-config-react-app": "^6.0.0",
    "eslint-plugin-import": "2.23.4",
    "eslint-plugin-jsx-a11y": "6.4.1",
    "eslint-plugin-react": "7.24.0",
    "eslint-plugin-react-hooks": "4.2.0",
    "jest-environment-jsdom-sixteen": "^2.0.0",
    "jest-styled-components": "^7.0.6",
    "prettier": "2.3.1"
  }
}
