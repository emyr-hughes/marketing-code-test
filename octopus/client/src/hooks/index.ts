export { default as useQueryGetProductById } from "./useQueryGetProductById";
