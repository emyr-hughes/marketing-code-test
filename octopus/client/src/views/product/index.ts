export {default as Product} from "./State"