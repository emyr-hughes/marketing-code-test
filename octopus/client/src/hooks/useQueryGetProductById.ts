import { gql, QueryResult, useQuery } from "@apollo/client";
import { APIProduct } from "../types";

export const GET_PRODUCT = gql`
  query getProductById($id: ID) {
    product(productId: $id) {
      id
      name
      power
      description
      price
      quantity
      brand
      weight
      height
      width
      length
      modelCode
      colour
      imgUrl
    }
  }
`;

const useQueryGetProductById = (id: string): QueryResult =>
  useQuery<{ product: APIProduct }>(GET_PRODUCT, {
    variables: {
      id,
    },
  });
export default useQueryGetProductById;