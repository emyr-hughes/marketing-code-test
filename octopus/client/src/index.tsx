import React from "react";
import ReactDOM from "react-dom";
import { createGlobalStyle } from "styled-components";
import { ApolloProvider, ApolloClient, InMemoryCache } from "@apollo/client";

import App from "./App";
import * as serviceWorker from "./serviceWorker";

const GlobalStyles = createGlobalStyle`
  body {
    background: #050F2A;
    margin: 0;
    font-family: 'Gotham SSm A','Gotham SSm B',helvetica,arial,sans-serif;
    font-size: calc(0.6em + 1vw);
    padding-top: 0.4em;
    color: white;
  }
`;

const client = new ApolloClient({
  uri: "http://127.0.0.1:8000/graphql",
  cache: new InMemoryCache(),
});

const ApolloApp = (AppComponent: () => JSX.Element) => (
  <body>
    <GlobalStyles />
    <ApolloProvider client={client}>
      <AppComponent />
    </ApolloProvider>
  </body>
);

ReactDOM.render(ApolloApp(App), document.getElementById("root"));

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister();
