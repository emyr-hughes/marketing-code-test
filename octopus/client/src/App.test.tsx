import { fireEvent, render } from '@testing-library/react';
import { MockedProvider } from "@apollo/client/testing";
import { GET_PRODUCT } from "./hooks/useQueryGetProductById";
import { Product } from "./views";
import 'jest-styled-components'
import React from "react";

const mocks = [
  {
    request: {
      query: GET_PRODUCT,
      variables: {
        id: "1",
      },
    },
    result: {
      data: {
        product: {
          id: "1",
          name: "Energy saving light bulb",
          power: "25W",
          description:
            "Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches on instantly, no wait around warm start and flicker free features make for a great all purpose bulb",
          price: 1299,
          quantity: 4,
          brand: "Philips",
          weight: 77,
          height: 12.6,
          width: 6.2,
          length: 6.2,
          modelCode: "E27 ES",
          colour: "Cool daylight",
          imgUrl: "https://i.ibb.co/2nzwxnQ/bulb.png",
        },
      },
    },
  },
];


test("renders with mock data", async () => {
  const {findByTestId, container} = render(
    <MockedProvider mocks={mocks}>
      <Product />
   </MockedProvider>
   )

  const title = await findByTestId("product-title")
  expect(title).toHaveTextContent("Energy saving light bulb")

  expect(container).toMatchSnapshot();

  const subtitle = await findByTestId("product-subtitle")
  expect(subtitle).toHaveTextContent("25W // Packet of 4")

  const pounds = await findByTestId("price-pounds")
  expect(pounds).toHaveTextContent("12")

  const pence = await findByTestId("price-pence")
  expect(pence).toHaveTextContent(".99")

  const brand = await findByTestId("product-specification-Brand-value")
  expect(brand).toHaveTextContent("Philips")

  const weight = await findByTestId("product-specification-Item weight-value")
  expect(weight).toHaveTextContent("77")

  const dimensions = await findByTestId("product-specification-Dimensions-value")
  expect(dimensions).toHaveTextContent("12.6x6.2x6.2")

  const model = await findByTestId("product-specification-Item model number-value")
  expect(model).toHaveTextContent("E27 ES")

  const colour = await findByTestId("product-specification-Colour-value")
  expect(colour).toHaveTextContent("Cool daylight")

})

test("should be able to increase and decrease product quantity", async () => {
  const {findByTestId} = render(
    <MockedProvider mocks={mocks}>
      <Product />
   </MockedProvider>
   )

  const increaseButton = await findByTestId("quantity-increase-button")
  const quantityAmount = await findByTestId("quantity-amount")
  const decreaseButton = await findByTestId("quantity-decrease-button")
  fireEvent.click(increaseButton)
  expect(quantityAmount).toHaveTextContent("2")
  fireEvent.click(increaseButton)
  expect(quantityAmount).toHaveTextContent("3")
  fireEvent.click(decreaseButton)
  expect(quantityAmount).toHaveTextContent("2")
  fireEvent.click(decreaseButton)
  expect(quantityAmount).toHaveTextContent("1")
});


test("should be able to add items to the basket", async () => {
  const addToCartMock = jest.fn()
  const {findByTestId} = render(
    <MockedProvider mocks={mocks}>
      <Product handleAddToCart={addToCartMock}/>
   </MockedProvider>
   )

  const addToCartButton = await findByTestId("product-add-to-cart-button")
 fireEvent.click(addToCartButton)
 expect(addToCartMock).toHaveBeenCalled()
});
