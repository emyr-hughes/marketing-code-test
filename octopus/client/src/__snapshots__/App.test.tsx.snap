// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with mock data 1`] = `
.c12 {
  height: 2em;
}

.c13 {
  font-size: 2.9em;
  font-weight: bold;
}

.c14 {
  line-height: 1.3em;
  font-size: 2em;
  font-weight: bold;
}

.c17 {
  background: #1A2A43;
  border-radius: 5px;
  height: 1.9em;
  width: 1.8em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.5em;
}

.c19 {
  background: #36598D;
  border-radius: 5px;
  height: 1.9em;
  width: 1.8em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.5em;
}

.c15 {
  margin-bottom: 0.1em;
  color: #36598d;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  font-size: 1.2em;
  font-weight: 400;
}

.c18 {
  font-size: 2.4em;
  font-weight: bold;
  margin: 0.2em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  height: 2em;
}

.c5 {
  height: 2em;
}

.c7 {
  padding-left: 1.7em;
  padding-right: 1.7em;
  padding-bottom: 2em;
  padding-top: 2em;
}

.c21 {
  font-size: 1.8em;
  font-weight: bold;
  margin-bottom: 0.8em;
}

.c8 {
  background: #01193b;
}

.c25 {
  background: #01193b;
  padding-left: 1em;
  padding-right: 1.7em;
  padding-bottom: 2em;
  padding-top: 2em;
}

.c26 {
  font-size: 0.9em;
  line-height: 1.9em;
  font-weight: 100;
  color: #8ea2c1;
}

.c23 {
  border-bottom: 2px solid #01193b;
  padding-top: 0.8em;
  padding-bottom: 0.8em;
}

.c20 {
  background: #fb1a7b;
  border-radius: 5px;
  height: 2em;
  margin-top: 0.5em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-weight: bold;
  font-size: 2.3em;
}

.c1 {
  margin-bottom: 10px;
  padding-left: 1em;
  padding-right: 1em;
}

.c9 {
  margin-top: -2.5em;
}

.c6 {
  width: 75%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  z-index: 1;
}

.c10 {
  margin-top: 0.5em;
  font-size: 2.9em;
  font-weight: bold;
}

.c22 {
  font-size: 1.1em;
  line-height: 1.4em;
}

.c24 {
  font-size: 1.2em;
  width: 50%;
}

.c11 {
  margin-top: 1em;
  color: #36598d;
  font-size: 1.5em;
  font-weight: 400;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2 c3"
      >
        <img
          alt="logo"
          class="c4"
          src="[object Object]"
        />
        <img
          alt="basket"
          class="c5"
          src="[object Object]"
        />
      </div>
    </div>
    <img
      class="c6"
      src="https://i.ibb.co/2nzwxnQ/bulb.png"
    />
    <div
      class="c0 c7 c8 c9"
    >
      <div
        class="c10"
        data-testid="product-title"
      >
        Energy saving light bulb
      </div>
      <div
        class="c11"
        data-testid="product-subtitle"
      >
        25W // Packet of 4
      </div>
    </div>
    <div
      class="c0 c7"
    >
      <div
        class="c0"
      >
        <div
          class="c2 c3"
        >
          <div
            class="c0"
          >
            <div
              class="c12"
            />
            <div
              class="c2"
            >
              <div
                class="c13"
                data-testid="price-pounds"
              >
                12
              </div>
              <div
                class="c14"
                data-testid="price-pence"
              >
                .
                99
              </div>
            </div>
          </div>
          <div
            class="c0"
          >
            <div
              class="c15"
            >
              QTY
            </div>
            <div
              class="c16"
            >
              <div
                class="c17"
                data-testid="quantity-decrease-button"
              >
                -
              </div>
              <div
                class="c18"
                data-testid="quantity-amount"
              >
                1
              </div>
              <div
                class="c19"
                data-testid="quantity-increase-button"
              >
                +
              </div>
            </div>
          </div>
        </div>
        <div
          class="c20"
          data-testid="product-add-to-cart-button"
        >
          Add to cart
        </div>
      </div>
    </div>
    <div
      class="c0 c7 c8"
    >
      <div
        class="c21"
      >
        Description
      </div>
      <div
        class="c22"
      >
        Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches on instantly, no wait around warm start and flicker free features make for a great all purpose bulb
      </div>
    </div>
    <div
      class="c0 c7"
    >
      <div
        class="c21"
      >
        Specifications
      </div>
      <div
        class="c2 c3 c23"
      >
        <div
          class="c24"
          data-testid="product-specification-Brand-property"
        >
          Brand
        </div>
        <div
          class="c24"
          data-testid="product-specification-Brand-value"
        >
          Philips
        </div>
      </div>
      <div
        class="c2 c3 c23"
      >
        <div
          class="c24"
          data-testid="product-specification-Item weight-property"
        >
          Item weight
        </div>
        <div
          class="c24"
          data-testid="product-specification-Item weight-value"
        >
          77
        </div>
      </div>
      <div
        class="c2 c3 c23"
      >
        <div
          class="c24"
          data-testid="product-specification-Dimensions-property"
        >
          Dimensions
        </div>
        <div
          class="c24"
          data-testid="product-specification-Dimensions-value"
        >
          12.6x6.2x6.2
        </div>
      </div>
      <div
        class="c2 c3 c23"
      >
        <div
          class="c24"
          data-testid="product-specification-Item model number-property"
        >
          Item model number
        </div>
        <div
          class="c24"
          data-testid="product-specification-Item model number-value"
        >
          E27 ES
        </div>
      </div>
      <div
        class="c2 c3 c23"
      >
        <div
          class="c24"
          data-testid="product-specification-Colour-property"
        >
          Colour
        </div>
        <div
          class="c24"
          data-testid="product-specification-Colour-value"
        >
          Cool daylight
        </div>
      </div>
    </div>
    <div
      class="c0 c25"
    >
      <div
        class="c26"
      >
        Octopus Energy Ltd is a company registered in England and Wales.
      </div>
      <div
        class="c26"
      >
        Registered Number: 09263424. Registered office: 33 Holborn, London, EC1N 2HT. Trading office: 20-24 Broadwich Street, London, W1F 8HT.
      </div>
    </div>
  </div>
</div>
`;
